### ISSUE TITLE
Add Gymnasium interface to Minetest for training RL agents

### ISSUE THREAD

**Issue by siboehm** [minetest/minetest#14069]
### Problem

Training Reinforcement learning agents on game environments is a pretty active area of research (e.g. [DeepMind's XLand](https://deepmind.google/discover/blog/generally-capable-agents-emerge-from-open-ended-play/), [OpenAIs VPT on Minecraft](https://openai.com/research/vpt)). There is currently no easy way to interface with Minetest for training such agents.

Why would this be cool?
- Basically, the ideal training environment would be: Complex enough to yield interesting agent behavior, easily modifiable, computationally efficient (training agents requires running many copies of the environment in parallel), multiplayer support (for agents to interact with agents, and humans to interact with agents).
- To my knowledge, there are no complex Open Source games that are suitable for training RL agents. Most Open Source games are too simple (e.g. the default environments in [Gymnasium](https://gymnasium.farama.org/)), so instead researchers rely on in-house custom built environments (like DM's XLand, which is based on Unity), or on interfacing with closed-source games (like Minecraft, through [minerl](https://github.com/minerllabs/minerl)) through a keyboard & video interface. Trying to deal with Minecrafts's closed source nature makes the environment brittle and hard to maintain.
- Minetest would work super well for this: It's easily extendable (via mods and more deeply through modifying the engine code), it's computationally efficient enough to run many copies in parallel while training (particularly, minetest could be modified to run headless without relying on xvfb like MineRL), it supports multiplayer (for agents to interact with other agents, and agents interacting with humans).

### Solutions

There are some easy-to-implement features that enable basic, single-agent training and some more complex features that are only necessary for multi-agent-multi-human server setups.

What's necessary for basic, single-agent training:
1. Remote Interface: There needs to be some way for the agent (running in a different process) to send keyboard commands and mouse movements to the game. Most likely this could be implemented via either RPCs, or using a message queue. The game needs to send back screenshots, as well as rewards (rewards are scalar values calculated from the game state. The reward function could be implemented as a mod). Ultimately, there needs to be a [Gymnasium](https://gymnasium.farama.org/) interface implemented in Python, but that could be moved into a separate repo, as long as the interface with minetest is well defined.
2. Synchronous mode: For easy single-agent training, the clockrate of the game needs to be configurable. There needs to be a way to pause the game in between receiving keystrokes from the agent. This is necessary to train agents that run slower than realtime.
3. Headless rendering: During RL agent training, the game will most likely run on a Linux server. Therefore, it needs to be able to run and provide screenshots without an attached display. Could be done by using xvfb, but a better solution would be a 'real headless' mode that still renders on GPU.

Additional features that are not necessary for basic RL training environments, but are nice to have:
- Easily accessible debugging information, e.g. about which menus are open or replays of past agent behaviour.
- The ability to easily record humans playing the game, e.g. capturing the video stream and keyboard presses.
- Synchronous multi-agent mode: If multiple agents are running in the same game, pausing the game while waiting for each agent to pick it's next action would be useful.

### Alternatives

Some of these features could be implemented without touching minetest code, but it'd be less computationally efficient and would not work cross-platform. E.g. Remote interface is doable by using OS APIs, headless is possible via xvfb.

### Additional context

EleutherAI did quite some work on modifying minetest for the [Alignment MineTest](https://www.eleuther.ai/projects/alignment-minetest) project on this fork: https://github.com/EleutherAI/minetest. They have an excellent blogpost about their minetester library [here](https://blog.eleuther.ai/minetester-intro/). I've read through most of their changes. Their repo is somewhat behind master now, and it needs some love before being suitable for merging.

If maintainers here are generally excited to make it easier to train RL agents on games based on minetest_engine, then I will start sending some PRs to implement these features, starting with the remote interface.

**Comment by HybridDog** [comment]
I think in addition to reinforcement learning, such an interface can be used for automatic end-to-end testing of Minetest mods.

**Comment by bowen-xu** [comment]
@siboehm Do you work on this? I think this feature is quite worthwile. Minetest can be a good environment for AI agents.

**Comment by siboehm** [comment]
There is a working version here: https://github.com/Astera-org/minetest and also at https://github.com/EleutherAI/minetest. I was hoping we would eventually merge this into the Repo, but it requires quite a lot of modifications and new dependencies.

**Comment by rubenwardy** [comment]
Hey, I think the usecase of training AI should be done on a Minetest fork rather than with upstream support

End-to-end testing would be a compelling use case, but I'd suggest a more general issue for that rather than suggesting Gymnasium specifically

**Comment by mikelma** [comment]
Hi!

I've recently released a Python library + Minetest fork that implements the Gymnasium interface. It's named [Craftium](https://github.com/mikelma/craftium) and is focused on providing a platform for researchers to create and modify RL environments and tasks. Although it implements many of the features mentioned in the issue it's still in early development and needs testing, but I think it's pretty usable! Online docs are also available [here](https://craftium.readthedocs.io). Testing, reporting issues, and submitting fixes are very welcome! Don't hesitate to contact me if you have any questions.

Thanks!🤗
